File: 496.next-greater-element-i/solution.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        mem = {}
        stack = []
        for val2 in nums2:
            while (stack) and (stack[-1] < val2):
                mem[stack.pop()] = val2
            stack.append(val2)
            mem[val2] = -1

        result = []
        for val1 in nums1:
            result.append(mem[val1])
        return result

File: 496.next-greater-element-i/solution.kt
import java.util.Stack
class Solution {
    fun nextGreaterElement(nums1: IntArray, nums2: IntArray): IntArray {
        var mem = HashMap<Int, Int>()
        var stack = Stack<Int>()

        for (val2 in nums2) {
            while ((stack.size > 0) && (stack.peek() < val2)) {
                mem[stack.pop()] = val2
            }
            mem[val2] = -1
            stack.push(val2)
        }

        var result = ArrayList<Int>()
        for (val1 in nums1) {
            result.add(mem[val1]!!)
        }
        return result.toIntArray()
    }
}
